### DISCUSSION TITLE
Error Class 'XX' should have [public, protected] no-arg constructor

### DISCUSSION THREAD

**Discussion by tobiascastaneda2000** [dds-jv/foro#404] [Consultas de persistencia] [answered]
Buenas tardes!

Tengo una consulta surgida del uso de JPA.
Al realizar las anotaciones de @Entity, en algunas clases aparecían subrayadas en rojo con el siguiente error:

`Class 'Miembro' should have [public, protected] no-arg constructor`

Por lo que averigüe este error aparece cuando la clase no tiene un constructor sin argumentos, por lo que al agregarlo de la siguiente manera:

public Miembro() {
    
  }

Se soluciona el error.

Sin embargo, esto provocaría que haya dos constructores al mismo tiempo: El 1° un constructor sin parámetros, y el 2° un constructor con dos parámetros que se usaban para la lógica de los requerimientos anteriores (en este caso, dos parámetros). Esta encrucijada se repite para otras clases. ¿Qué me recomienda hacer?

¿Puedo dejar dos constructores, uno vacío Miembro() y otro con parámetros Miembro(parametro1, parametro2, etc)?

¿Por qué me exige tener un constructor sin parámetros?

Disculpe las molestias, espero su respuesta.

**Comment by flbulgarelli** [comment]
¡Buena pregunta! El motivo es que JPA instancia a los objetos persistentes usando _reflection_ (es decir, una técnica de metaprogramación) y por tanto no sabe qué argumentos pasarle al constructor. Es por ello que necesita un constructor fácil de invocar: público y sin argumentos.

**Reply by flbulgarelli** [reply, marked as answer]
Dicho eso, sí, vas a tener que dejar ambos constructores: uno público (o mejor, _protected_, para minimizar la probabilidad de errores y documentar mejor tu intención) y otro, con los parámetros que ya tenés.
